File: generated_code/cpp/solution_3772_1.cpp
#include <vector>
#include <algorithm>
#include <set>
#include <numeric>

using namespace std;

class Solution {
public:
    int minimumPairRemoval(vector<int>& nums) {
        int n = nums.size();
        if (n <= 1) return 0;

        int ans = 0;
        int inversionsCount = 0;

        // Doubly linked list tracking for current active elements
        vector<int> nextIndices(n);
        vector<int> prevIndices(n);
        vector<long long> values(nums.begin(), nums.end());

        for (int i = 0; i < n; ++i) {
            nextIndices[i] = i + 1;
            prevIndices[i] = i - 1;
        }

        // Custom comparator for the set to ensure we pick the minimum sum pair.
        // If sums are equal, we break ties by selecting the leftmost pair (smaller index).
        auto comp = [&](const pair<long long, int>& a, const pair<long long, int>& b) {
            if (a.first != b.first) return a.first < b.first;
            return a.second < b.second;
        };
        set<pair<long long, int>, decltype(comp)> pairSums(comp);

        // Populate initial adjacent pair sums and inversion counts
        for (int i = 0; i < n - 1; ++i) {
            pairSums.insert({values[i] + values[i + 1], i});
            if (values[i + 1] < values[i]) {
                ++inversionsCount;
            }
        }

        // Loop dynamically merges pairs as long as the array is not fully non-decreasing
        while (inversionsCount > 0) {
            ++ans;
            
            auto smallestPair = *pairSums.begin();
            pairSums.erase(pairSums.begin());

            long long pairSum = smallestPair.first;
            int currIndex = smallestPair.second;

            int nextIndex = nextIndices[currIndex];
            int prevIndex = prevIndices[currIndex];

            // Re-evaluate inversion changes around the current left element before it absorbs nextIndex
            if (prevIndex >= 0) {
                long long oldPairSum = values[prevIndex] + values[currIndex];
                long long newPairSum = values[prevIndex] + pairSum;
                pairSums.erase({oldPairSum, prevIndex});
                pairSums.insert({newPairSum, prevIndex});

                if (values[prevIndex] > values[currIndex]) --inversionsCount;
                if (values[prevIndex] > pairSum) ++inversionsCount;
            }

            if (values[nextIndex] < values[currIndex]) --inversionsCount;

            // Re-evaluate inversion changes around the element next to nextIndex
            int nextNextIndex = (nextIndex < n) ? nextIndices[nextIndex] : n;
            if (nextNextIndex < n) {
                long long oldPairSum = values[nextIndex] + values[nextNextIndex];
                long long newPairSum = pairSum + values[nextNextIndex];
                pairSums.erase({oldPairSum, nextIndex});
                pairSums.insert({newPairSum, currIndex});

                if (values[nextNextIndex] < values[nextIndex]) --inversionsCount;
                if (values[nextNextIndex] < pairSum) ++inversionsCount;
                
                prevIndices[nextNextIndex] = currIndex;
            }

            // Perform the logical merge in the doubly linked list structures
            nextIndices[currIndex] = nextNextIndex;
            values[currIndex] = pairSum;
        }

        return ans;
    }
};

File: generated_code/python/solution_3772_1.py
from typing import List
from sortedcontainers import SortedList

class Solution:
    def minimumPairRemoval(self, nums: List[int]) -> int:
        n = len(nums)
        ans = 0
        
        # Track the number of inversions (where nums[i] > nums[i+1])
        # The array is sorted / non-decreasing when inversionsCount == 0
        inversionsCount = sum(nums[i + 1] < nums[i] for i in range(n - 1))
        
        # Doubly linked list representations using indexing
        nextIndices = [i + 1 for i in range(n)]
        prevIndices = [i - 1 for i in range(n)]
        
        # Sorted list acting as a min-heap to keep track of adjacent pair sums.
        # Elements are stored as (pair_sum, left_index). Ties are naturally broken
        # by the leftmost index due to the tuple comparison format.
        pairSums = SortedList((nums[i] + nums[i + 1], i) for i in range(n - 1))
        
        while inversionsCount > 0:
            ans += 1
            # Retrieve and remove the leftmost pair with the minimum sum
            smallestPair = pairSums.pop(0)
            pairSum, currIndex = smallestPair
            
            nextIndex = nextIndices[currIndex]
            prevIndex = prevIndices[currIndex]
            
            # If a previous element exists, the relationship between prevIndex and currIndex changes
            if prevIndex >= 0:
                oldPairSum = nums[prevIndex] + nums[currIndex]
                newPairSum = nums[prevIndex] + pairSum
                pairSums.remove((oldPairSum, prevIndex))
                pairSums.add((newPairSum, prevIndex))
                
                # Update inversion status for (prevIndex, currIndex)
                if nums[prevIndex] > nums[currIndex]:
                    inversionsCount -= 1
                if nums[prevIndex] > pairSum:
                    inversionsCount += 1
                    
            # Update inversion status for the merged pair (currIndex, nextIndex) being removed
            if nums[nextIndex] < nums[currIndex]:
                inversionsCount -= 1
                
            nextNextIndex = nextIndices[nextIndex] if nextIndex < n else n
            
            # If an element exists after the merged pair, update its relationship with currIndex
            if nextNextIndex < n:
                oldPairSum = nums[nextIndex] + nums[nextNextIndex]
                newPairSum = pairSum + nums[nextNextIndex]
                pairSums.remove((oldPairSum, nextIndex))
                pairSums.add((newPairSum, currIndex))
                
                # Update inversion status for (nextIndex, nextNextIndex)
                if nums[nextNextIndex] < nums[nextIndex]:
                    inversionsCount -= 1
                if nums[nextNextIndex] < pairSum:
                    inversionsCount += 1
                    
                prevIndices[nextNextIndex] = currIndex
                
            # Finalize the removal of nextIndex and update the value of currIndex
            nextIndices[currIndex] = nextNextIndex
            nums[currIndex] = pairSum
            
        return ans